external_plugins/custom_tokens/custom_tokens_plugin.cpp

```diff
@@ -0,0 +1,367 @@
+         token_sequence = itr->token_sequence + 1;
+   }
+
+   {
+      const auto &idx = db_->get_index<token_operation_index>().indices().get<by_account>();
+      auto itr = idx.lower_bound(boost::make_tuple(account, uint64_t(-1)));
+      if( itr != idx.end() && itr->account == account )
+         account_sequence = itr->account_sequence + 1;
+   }
+
+   db_->create<custom_token_operation_object>([ & ](custom_token_operation_object &to) {
+        to.account = account;
+        to.token_symbol = token_symbol;
+        to.trx_id = note.trx_id;
+        to.block = note.block;
+        to.timestamp = db_->head_block_time();
+        to.account_sequence = account_sequence;
+        to.token_sequence = token_sequence;
+        auto size = fc::raw::pack_size(note.op);
+        to.serialized_op.resize(size);
+        fc::datastream<char *> ds(to.serialized_op.data(), size);
+        fc::raw::pack(ds, note.op);
+   });
+}
+
+const std::string custom_tokens_plugin_impl::save_token_error(const transaction_id_type &tx_id,
+                                                              const std::string &error,
+                                                              asset_symbol_type token_symbol) {
+
+   uint64_t token_sequence = 1;
+   {
+      const auto &idx = db_->get_index<token_error_index>().indices().get<by_token_symbol>();
+      auto itr = idx.lower_bound(boost::make_tuple(token_symbol, uint64_t(-1)));
+      if( itr != idx.end() && itr->token_symbol == token_symbol )
+         token_sequence = itr->token_sequence + 1;
+   }
+
+   db_->create<custom_token_error_object>([ & ](custom_token_error_object &cto) {
+        cto.token_symbol = token_symbol;
+        cto.trx_id = tx_id;
+        cto.token_sequence = token_sequence;
+        from_string(cto.error, error);
+   });
+
+   if( prune_ ) {
+      // Clean up errors older then 30 days
+      auto now = std::chrono::system_clock::now();
+      const auto &seq_idx = db_->get_index<token_error_index, by_time>();
+      auto seq_itr = seq_idx.lower_bound(now);
+      vector<const custom_token_error_object *> to_remove;
+
+
+      if( seq_itr == seq_idx.begin())
+         return error;
+
+      --seq_itr;
+
+      while( now - seq_itr->time > std::chrono::hours(24 * 30)) {
+         to_remove.push_back(&(*seq_itr));
+         --seq_itr;
+      }
+
+      for( const auto *seq_ptr : to_remove ) {
+         db_->remove(*seq_ptr);
+      }
+   }
+   return error;
+}
+
+} // detail
+
+custom_tokens_plugin::custom_tokens_plugin() {}
+
+void custom_tokens_plugin::set_program_options(options_description &cli, options_description &cfg) {
+   cfg.add_options()
+         ("custom-token-app-id", boost::program_options::value<uint64_t>()->default_value(2),
+          "App id used by the custom token plugin")
+         ("custom-token-error-pruning", boost::program_options::value<bool>()->default_value(true),
+          "Enable/Disable cleaning (more then 30 days) old custom token errors");
```

Comment:
then/than typo... this param might specify max. time-frame they want to keep errors for instead of enable -> fixed 30 days time-frame or disable -> infinite time...